subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultTaskOutputFilesRepository.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import org.gradle.api.invocation.Gradle;
+import org.gradle.cache.CacheBuilder;
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.PersistentCache;
+import org.gradle.cache.PersistentIndexedCache;
+import org.gradle.cache.PersistentIndexedCacheParameters;
+import org.gradle.cache.internal.FileLockManager;
+import org.gradle.internal.nativeintegration.filesystem.FileType;
+import org.gradle.util.GradleVersion;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
+public class DefaultTaskOutputFilesRepository implements TaskOutputFilesRepository, Closeable {
+
+    private final static String CACHE_DISPLAY_NAME = "Build Output Cleanup Cache";
+
+    private final CacheRepository cacheRepository;
+    private final PersistentCache cacheAccess;
+    private final FileSystemMirror fileSystemMirror;
+    private final PersistentIndexedCache<String, Boolean> outputFiles;
+    private final PersistentIndexedCache<String, Boolean> outputFileParents;
+
+    public DefaultTaskOutputFilesRepository(CacheRepository cacheRepository, Gradle gradle, FileSystemMirror fileSystemMirror, InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory) {
+        this.cacheRepository = cacheRepository;
+        this.cacheAccess = createCache(gradle);
+        this.fileSystemMirror = fileSystemMirror;
+        this.outputFiles = cacheAccess.createCache(cacheParameters("outputFiles", inMemoryCacheDecoratorFactory));
+        this.outputFileParents = cacheAccess.createCache(cacheParameters("outputFileParents", inMemoryCacheDecoratorFactory));
+    }
+
+    private static PersistentIndexedCacheParameters<String, Boolean> cacheParameters(String cacheName, InMemoryCacheDecoratorFactory inMemoryCacheDecoratorFactory) {
+        return new PersistentIndexedCacheParameters<String, Boolean>(cacheName, String.class, Boolean.class)
+            .cacheDecorator(inMemoryCacheDecoratorFactory.decorator(100000, true));
```

Comment:
Gut feeling? More scientific: The Gradle sanity check build has around 2500 tasks. We need to track the outputs of those tasks. Most task have one or two output properties. So we end up with 5000 output directories which we need to track. Assuming that each path contains about 10 segments we would need to keep 50000 elements for the whole build in memory. Now let's aim for twice the size of the Gradle build, so we end up with 100000.
Do you have a better number? Or do you have an experiment which we could try to find out what a good value could be?